File: server/pkg/domain/game.go
package domain

import "time"

type Game struct {
	ID                  string                `json:"id"`
	Admin               *User                 `json:"admin"`
	State               GameState             `json:"state"`
	Route               *Route                `json:"route,omitempty"`
	Players             []*Player             `json:"players"`
	ActivePoll          *Poll                 `json:"activePoll,omitempty"`
	CompletedTaskPoints []*CompletedTaskPoint `json:"completedTaskPoints"`

	CreatedAt  time.Time  `json:"createdAt"`
	FinishedAt *time.Time `json:"finishedAt"`
}

type GameState string

type CreateGame struct {
	AdminID string    `json:"admin_id"`
	State   GameState `json:"state"`
}

type PatchGame struct {
	State      *GameState  `json:"state"`
	RouteID    *string     `json:"routeId"`
	FinishedAt **time.Time `json:"finishedAt"`
}

type FilterGame struct {
	ID       *string    `json:"id"`
	AdminID  *string    `json:"adminId"`
	PlayerID *string    `json:"playerId"`
	State    *GameState `json:"state"`
	Limit    *int       `json:"limit"`

	SortByCreatedAtDesc bool `json:"sortByCreatedAtDesc"`
	IncludeAdmin        bool `json:"includeAdmin"`
	IncludeRoute        bool `json:"includeRoute"`
}

const (
	GameStateLobby    GameState = "lobby"
	GameStateGame     GameState = "game"
	GameStatePoll     GameState = "poll"
	GameStateFinished GameState = "finished"
)

File: client/src/shared/interfaces/game/game.ts
import { GameState } from "./game-state";
import { Player } from "./player";
import { User } from "../user";
import { Route } from "../route";
import { Poll } from "../polls/poll";
import { CompletedTaskPoint } from "./completed-task-point";

export interface Game {
  id: string;
  admin: User;
  createdAt: string;
  finishedAt: string;
  players: Player[];
  state: GameState;
  activePoll: Poll | null;
  route: Route | null; // we know that these will not be null if game is not lobby
  completedTaskPoints: CompletedTaskPoint[];
}
